### ISSUE TITLE
API for go repository

### ISSUE THREAD

**Issue by g4s8** [artipie/go-adapter#20]
Let's add API (`Slice` implementation from https://github.com/artipie/http) to provide HTTP API for Golang repository.
Let's start with documentation and describe how does it work, then create integration test, then implement `Slice` API.

Documentation should be added to the `README.md` file, it should describe how does go packages be downloaded and uploaded, what API methods and endpoints it can use and how.

Integration test can be implemented using Docker image with `go` CLI (using `go get` to download packages), [Testcontainers](https://github.com/testcontainers/testcontainers-java) can be used to integrate it with Java code (see integration tests in `artipie/rpm-adapter` for reference).

`Slice` implementation should encapsulate `Storage` from `artipie/asto` where all go packages are stored. See `artipie/files-adapter` and `artipie/npm-adapter` for existing `Slice` implementations.

**Comment by 0crat** [comment]
The job [#20](https://github.com/artipie/go-adapter/issues/20) assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/6644c4e5-5cbd-47ca-843d-63f4e9120466); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by carlosmiranda** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/go-adapter/issues/20#issuecomment-605957660))

@carlosmiranda The user @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) resigned from [#20](https://github.com/artipie/go-adapter/issues/20), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/go-adapter/issues/20#issuecomment-605957660))

@carlosmiranda Job refused in 0 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
The job [#20](https://github.com/artipie/go-adapter/issues/20) assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/c70d3a83-251b-4f91-a9f9-3e8ad2a70f3f); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
@g4s8[/z](https://www.0crat.com/u/g4s8) I see that you unassigned this issue; the order is still assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
The job [#20](https://github.com/artipie/go-adapter/issues/20) assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/b829e863-ddea-4b9a-9e43-0cefe6f6728c); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); we should be aware that @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova) is on vacation; this ticket may be delayed; there will be no monetary reward for this job

**Comment by olenagerasimova** [comment]
@g4s8 here is what I discovered investigating how `go get` works:
1. without go proxy protocol `go get` works only with vcs, supported vcs are Bazaar, Fossil, Git, Mercurial, SVN. How it actually resolves and searches sources is described in [this](https://golang.org/cmd/go/#hdr-Remote_import_paths) section about importing: when we do `go get` not from well known code hosting resource, smth like 
```
go get example.org/pkg/foo
```
go tool attempts to fetch import data from meta-tag of the following form
```
<meta name="go-import" content="import-prefix vcs repo-root">
```
requesting it by the address
```
http://example.org/pkg/foo?go-get=1
```
After getting required info from the meta-tag, for example
```
<meta name="go-import" content="example.org git https://code.org/r/p/exproj">
```
`go get` will download code using specified vcs.

2. For modules vcs is not required and vcs name in meta-tag can be replaced with `mod`:
```
<meta name="go-import" content="example.org mod https://code.org/moduleproxy">
``` 
which means to fetch module from proxy available at the URL https://code.org/moduleproxy using go module proxy protocol.

So, to summer it up  - `go get` works either with vcs (not necessarily public, for example, it can be gitlab configured by a compony and available only inside their network) or over proxy.

**Comment by olenagerasimova** [comment]
@g4s8 let me add some more information: 
1. While using Artifactory there are [2 options](https://jfrog.com/blog/how-to-use-jfrog-artifactory-and-gocenter-together-to-build-go-apps/) to get go modules: the first one is proxy and the second one to use [JFrog CLI](https://www.jfrog.com/confluence/display/CLI/JFrog+CLI) 
2. As go [module proxy protocol](https://golang.org/cmd/go/#hdr-Module_proxy_protocol) requests are get requests without parameters, it's enough to expose file system/folder/catalog over http, also `file:///` urls are supported.

**Comment by g4s8** [comment]
@olenagerasimova as I understand correctly `go get` download from git using https by default, e.g. downloading from github via `git://` protocol requires authentication, but `go get` works fine without it.

**Comment by olenagerasimova** [comment]
@g4s8 to get smth from git `go get` uses git, you will not be able to install go module from git without git installed, authorisation info is stored in your git settings. Run `go get` with `-x` flag to get all the details, you can try not too long example from [this](https://nanxiao.gitbooks.io/golang-101-hacks/content/posts/go-get-command.html) article. In logs you can see that go either use git or performs get requests using proxy protocol.

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#24](https://github.com/artipie/go-adapter/issues/24) is still not solved.

**Comment by 0crat** [comment]
@olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by olenagerasimova** [comment]
@g4s8 please close this issue, PR #23 was merged

**Comment by g4s8** [comment]
@olenagerasimova thanks

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#30](https://github.com/artipie/go-adapter/issues/30) is still not solved; solved: [#24](https://github.com/artipie/go-adapter/issues/24).

**Comment by 0pdd** [comment]
@g4s8 6 puzzles [#30](https://github.com/artipie/go-adapter/issues/30), [#31](https://github.com/artipie/go-adapter/issues/31), [#32](https://github.com/artipie/go-adapter/issues/32), [#33](https://github.com/artipie/go-adapter/issues/33), [#34](https://github.com/artipie/go-adapter/issues/34), [#35](https://github.com/artipie/go-adapter/issues/35) are still not solved; solved: [#24](https://github.com/artipie/go-adapter/issues/24).

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#31](https://github.com/artipie/go-adapter/issues/31), [#32](https://github.com/artipie/go-adapter/issues/32), [#33](https://github.com/artipie/go-adapter/issues/33), [#35](https://github.com/artipie/go-adapter/issues/35) are still not solved; solved: [#24](https://github.com/artipie/go-adapter/issues/24), [#30](https://github.com/artipie/go-adapter/issues/30), [#34](https://github.com/artipie/go-adapter/issues/34).

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#31](https://github.com/artipie/go-adapter/issues/31), [#32](https://github.com/artipie/go-adapter/issues/32), [#35](https://github.com/artipie/go-adapter/issues/35), [#39](https://github.com/artipie/go-adapter/issues/39) are still not solved; solved: [#24](https://github.com/artipie/go-adapter/issues/24), [#30](https://github.com/artipie/go-adapter/issues/30), [#33](https://github.com/artipie/go-adapter/issues/33), [#34](https://github.com/artipie/go-adapter/issues/34).

**Comment by 0pdd** [comment]
@g4s8 3 puzzles [#31](https://github.com/artipie/go-adapter/issues/31), [#32](https://github.com/artipie/go-adapter/issues/32), [#35](https://github.com/artipie/go-adapter/issues/35) are still not solved; solved: [#24](https://github.com/artipie/go-adapter/issues/24), [#30](https://github.com/artipie/go-adapter/issues/30), [#33](https://github.com/artipie/go-adapter/issues/33), [#34](https://github.com/artipie/go-adapter/issues/34), [#39](https://github.com/artipie/go-adapter/issues/39).

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#42](https://github.com/artipie/go-adapter/issues/42) is still not solved; solved: [#24](https://github.com/artipie/go-adapter/issues/24), [#30](https://github.com/artipie/go-adapter/issues/30), [#31](https://github.com/artipie/go-adapter/issues/31), [#32](https://github.com/artipie/go-adapter/issues/32), [#33](https://github.com/artipie/go-adapter/issues/33), [#34](https://github.com/artipie/go-adapter/issues/34), [#35](https://github.com/artipie/go-adapter/issues/35), [#39](https://github.com/artipie/go-adapter/issues/39).

**Comment by 0pdd** [comment]
@g4s8 all 9 puzzles are solved here: [#24](https://github.com/artipie/go-adapter/issues/24), [#30](https://github.com/artipie/go-adapter/issues/30), [#31](https://github.com/artipie/go-adapter/issues/31), [#32](https://github.com/artipie/go-adapter/issues/32), [#33](https://github.com/artipie/go-adapter/issues/33), [#34](https://github.com/artipie/go-adapter/issues/34), [#35](https://github.com/artipie/go-adapter/issues/35), [#39](https://github.com/artipie/go-adapter/issues/39), [#42](https://github.com/artipie/go-adapter/issues/42).
